### PR TITLE
Improve RULES situation

### LINKED ISSUE #4: -Wall cleanliness: RULE pragma warnings
I disable this warning in the .cabal as it is probably harmless. Should be fixed properly at some point.

### PR DESCRIPTION
**Pull request by treeowl**
Add staging to make rules more likely to work out properly
and get rid of rule/inline warnings.

Fixes #4

### CODE REVIEW

File: src/Data/PQueue/Min.hs [hunk 1]
```diff
@@ -248,17 +259,19 @@ toList :: Ord a => MinQueue a -> [a]
 toList = toAscList
 
 {-# RULES
-  "toAscList" forall q . toAscList q = build (\c nil -> foldrAsc c nil q);
-    -- inlining doesn't seem to be working out =/
-  "toDescList" forall q . toDescList q = build (\c nil -> foldrDesc c nil q);
-  #-}
+"toAscList" [~1] forall q . toAscList q = build (\c nil -> foldrAsc c nil q)
+"toDescList" [~1] forall q . toDescList q = build (\c nil -> foldrDesc c nil q)
```
Comment (konsumlamm) [inline src/Data/PQueue/Min.hs:263]:
```suggestion
"toAscList" [~1] forall q. toAscList q = build (\c nil -> foldrAsc c nil q)
"toDescList" [~1] forall q. toDescList q = build (\c nil -> foldrDesc c nil q)
```

Comment (konsumlamm) [inline src/Data/PQueue/Min.hs:263]:
```suggestion
"toAscList" [~1] forall q. toAscList q = build (\c nil -> foldrAsc c nil q)
"toDescList" [~1] forall q. toDescList q = build (\c nil -> foldrDesc c nil q)
```

**Review by konsumlamm — CHANGES_REQUESTED** [review]
I've never used staging or rewrite rules, so I can't comment on that. I don't really like rewrite rules, but this just modifies existing ones, so I have no problem.

I see you've added `Data.PQueue.Min.foldlAsc'`, that should be added to the changelog. Is there any reason you didn't add `foldrAsc'`, `foldlDesc'` etc? I don't see any obvious reason why `foldlAsc'` should be special.

**Comment by treeowl** [comment]
> I see you've added `Data.PQueue.Min.foldlAsc'`, that should be added to the changelog. Is there any reason you didn't add `foldrAsc'`, `foldlDesc'` etc? I don't see any obvious reason why `foldlAsc'` should be special.

For min-queues, the natural/efficient order-respecting folds are `foldrAsc`, `foldlAsc'`, `foldlDesc`, and `foldrDesc'`. I'm not convinced we want the wonky ones we already have (`foldlAsc` and `foldrDesc`) let alone any new ones. But you're right; I should deal with that all in a separate PR.

**Pull request merged**